File: solutions/ruby/parallel-letter-frequency/1/parallel_letter_frequency.rb
# Write your code for the 'Parallel Letter Frequency' exercise in this file. Make
# the tests in `parallel_letter_frequency_test.rb` pass.
#
# To get started with TDD, see the `README.md` file in your
# `ruby/parallel_letter_frequency` directory.

class ParallelLetterFrequency
  def self.count(texts)
    return {} if texts.empty?

    threads = []
    texts.each do |text|
      threads << Thread.new do
        clean_text(text).split('').each_with_object(Hash.new(0)) do |char, hash|
          hash[char] += 1
        end
      end
    end

    threads.each(&:join)

    threads.map(&:value).reduce do |result, hash|
      result.merge(hash) { |_key, oldval, newval| oldval + newval }
    end
  end

  def self.clean_text(text)
    text.downcase.gsub(/[^\p{L}]/, '')
  end
end

File: solutions/go/parallel-letter-frequency/1/parallel_letter_frequency.go
// Dedicated to Shree DR.MDD
package letter

type FreqMap map[rune]int

func Frequency(text string) FreqMap {
	counts := FreqMap{}
	for _, char := range text {
		counts[char]++
	}
	return counts
}

func ConcurrentFrequency(lines []string) FreqMap {
	ch := make(chan FreqMap)
	for _, l := range lines {
		go func(s string) {
			ch <- Frequency(s)
		}(l)
	}
	result := FreqMap{}
	for range lines {
		for r, n := range <-ch {
			result[r] += n
		}
	}
	return result
}
